sdk/all/src/main/java/io/opentelemetry/sdk/OpenTelemetrySdk.java

```diff
@@ -5,50 +5,351 @@
+      this.resource = resource;
+      return this;
+    }
+
+    /**
+     * Returns a new {@link OpenTelemetrySdk} built with the configuration of this {@link Builder}.
+     */
+    @Override
+    public OpenTelemetrySdk build() {
+      BaggageManager baggageManager = this.baggageManager;
+      if (baggageManager == null) {
+        BaggageManagerFactory baggageManagerFactory = loadSpi(BaggageManagerFactory.class);
+        if (baggageManagerFactory != null) {
+          baggageManager = baggageManagerFactory.create();
+        } else if (HAS_BAGGAGE_SDK) {
+          baggageManager = new BaggageManagerSdk();
+        } else {
+          baggageManager = DefaultBaggageManager.getInstance();
+        }
+      }
+
+      MeterProvider meterProvider = this.meterProvider;
+      if (meterProvider == null) {
+        MeterProviderFactory meterProviderFactory = loadSpi(MeterProviderFactory.class);
+        if (meterProviderFactory != null) {
+          meterProvider = meterProviderFactory.create();
+        } else if (HAS_METRICS_SDK) {
+          meterProvider = MeterSdkProvider.builder().setClock(clock).setResource(resource).build();
+        } else {
+          meterProvider = DefaultMeterProvider.getInstance();
+        }
+      }
+
+      TracerProvider tracerProvider = this.tracerProvider;
+      if (tracerProvider == null) {
+        TracerProviderFactory tracerProviderFactory = loadSpi(TracerProviderFactory.class);
+        if (tracerProviderFactory != null) {
+          tracerProvider = new ObfuscatedTracerProvider(tracerProviderFactory.create());
+        } else if (HAS_TRACING_SDK) {
+          tracerProvider =
+              new ObfuscatedTracerProvider(
+                  TracerSdkProvider.builder().setClock(clock).setResource(resource).build());
+        } else {
+          tracerProvider = DefaultTracerProvider.getInstance();
+        }
+      }
+
+      OpenTelemetrySdk sdk =
+          new OpenTelemetrySdk(
+              tracerProvider, meterProvider, baggageManager, propagators, clock, resource);
+      // Automatically initialize global OpenTelemetry with the first SDK we build.
+      if (INITIALIZED_GLOBAL.compareAndSet(/* expectedValue= */ false, /* newValue= */ true)) {
+        OpenTelemetry.set(sdk);
+      }
+      return sdk;
+    }
   }
 
   /**
-   * Returns a {@link MeterSdkProvider}.
+   * Load provider class via {@link ServiceLoader}. A specific provider class can be requested via
+   * setting a system property with FQCN.
    *
-   * @return MeterProvider returned by {@link OpenTelemetry#getMeterProvider()}.
+   * @param providerClass a provider class
+   * @param <T> provider type
+   * @return a provider or null if not found
+   * @throws IllegalStateException if a specified provider is not found
    */
-  public static MeterSdkProvider getMeterProvider() {
-    return (MeterSdkProvider) OpenTelemetry.getMeterProvider();
+  @Nullable
+  private static <T> T loadSpi(Class<T> providerClass) {
```

Comment:
It bums me out that this method is an exact copy of the one in DefaultOpenTelemetry. It seems like having a common SpiUtilities somewhere in the API would be useful so we don't have to repeat ourselves on this boilerplate.

Reply:
Maybe add an abstract `AbstractOpenTelemetry` class to the api that has this method, and maybe others, that `DefaultOpenTelemetry` and `OpenTelemetrySdk` can both extend?

Reply:
Yeah it's not ideal but since it's private and still relatively small, I'm hoping to address this in a followup instead of adding more public API right away for this.